src/mame/drivers/dfs500.cpp

```diff
@@ -0,0 +1,1036 @@
+	uint8_t m_sel10;
+	uint8_t m_sel32;
+	uint8_t m_sel54;
+	uint8_t m_input_sel_A;
+	uint8_t m_input_sel_B;
+	uint16_t m_maincpu_latch16;
+	uint16_t m_effectcpu_latch16;
+	bool m_TOC;
+	bool m_TOE;
+	// TODO: bool m_BVS;
+	uint8_t m_trans_rate[2];
+	uint8_t m_pattern_number[3];
+	uint8_t m_userprogram_status;
+	uint8_t m_userprogram_edit;
+	uint8_t m_trail_duration;
+	uint8_t m_snapshot;
+	uint8_t m_dot_points1;
+	uint8_t m_dot_points2;
+	bool m_buzzer_state;
+
+	required_device<v20_device> m_cpanelcpu;
+	required_device<v30_device> m_maincpu;
+	required_device<v30_device> m_effectcpu;
+	required_device<pit8254_device> m_pit;
+	required_device<pic8259_device> m_pic;
+	required_device<i8251_device> m_serial1;
+	required_device<i8251_device> m_serial2;
+	required_device<i8251_device> m_cpanel_serial;
+	required_device<pit8254_device> m_cpanel_pit;
+	required_device<upd7004_device> m_adc;
+	required_memory_bank m_rombank1;
+	required_memory_bank m_rombank2;
+	required_device<beep_device> m_buzzer;
+	required_device<screen_device> m_screen;
+	required_device<picture_image_device> m_input1;
+	required_device<picture_image_device> m_input2;
+	required_device<picture_image_device> m_input3;
+	required_device<picture_image_device> m_input4;
+};
+
+IRQ_CALLBACK_MEMBER(dfs500_state::irq_callback)
+{
+	return m_int_vector;
+}
+
+void dfs500_state::machine_start()
+{
+	m_buzzer_state = false;
+	m_buzzer->set_state(0);
+	m_serial1->write_cts(0);
+	m_serial2->write_cts(0);
+	m_cpanel_serial->write_cts(0);
+	m_rombank1->configure_entries(0, 128, memregion("effectdata")->base(), 0x4000);
+	m_rombank2->configure_entries(0, 128, memregion("effectdata")->base(), 0x4000);
+	m_rombank1->set_entry(0);
+	m_rombank2->set_entry(0);
+	m_maincpu_latch16 = 0x0000;
+	m_effectcpu_latch16 = 0x0000;
+	m_TOC = false;
+	m_TOE = false;
+	m_int_vector = 0x00;
+	m_sel10 = 0;
+	m_sel32 = 0;
+	m_sel54 = 0;
+}
+
+u32 dfs500_state::screen_update(screen_device &screen, bitmap_rgb32 &bitmap, rectangle const &cliprect)
+{
+	bitmap_argb32 *input_bitmap;
+	switch(m_input_sel_A & 3)
+	{
+		case 0: input_bitmap = &m_input1->get_bitmap(); break;
+		case 1: input_bitmap = &m_input2->get_bitmap(); break;
+		case 2: input_bitmap = &m_input3->get_bitmap(); break;
+		case 3: input_bitmap = &m_input4->get_bitmap(); break;
+	}
+
+	if (input_bitmap)
+	{
+		// convert arbitrary sized ARGB32 image to a full-screen image
+		double stepx = (double)input_bitmap->width() / VIDEO_WIDTH;
+		double stepy = (double)input_bitmap->height() / VIDEO_HEIGHT;
+
+		for (unsigned screen_y = screen.visible_area().min_y; screen_y <= screen.visible_area().max_y; screen_y++)
+		{
+			for (unsigned screen_x = screen.visible_area().min_x; screen_x <= screen.visible_area().max_x; screen_x++)
+			{
+				bitmap.pix(screen_y, screen_x) = input_bitmap->pix((int)((double)(screen_y % VIDEO_HEIGHT) * stepy),
+				                                                   (int)((double)(screen_x % VIDEO_WIDTH) * stepx));
```

Comment:
Don’t try aligning stuff like this – it displays incorrectly when people use different tab sizes.  Just use two tabs for line continuations.

Reply:
The second line contains tabs up until the letter b of `bitmap` of the first one and then spaces for the intended alignment. It will work in all possible text editor setups.

Reply:
You can see that we run an automated whitespace normalisation tool on the source before forking each release.  Indentation outside comments is converted to tabs.  The kind of formatting you’ve used here will not be preserved.

If you need to break up the arguments across lines, indent them two tabs from the level of the call so it’s clear it’s a statement continuation and not a new scope level.  Also, lose some of the superfluous parentheses.
```
				bitmap.pix(screen_y, screen_x) = input_bitmap->pix(
						int(double(screen_y % VIDEO_HEIGHT) * stepy),
						int(double(screen_x % VIDEO_WIDTH) * stepx));
```

Reply:
done